xdsl/irdl/attributes.py

```diff
@@ -172,15 +169,54 @@ def from_pyrdl(
             if key not in _PARAMETRIZED_ATTRIBUTE_DICT_KEYS
         }
 
-        # Check that all fields of the attribute definition are either already
-        # in ParametrizedAttribute, or are class functions or methods.
+        # The resulting parameters
+        parameters: dict[str, AttrConstraint] = {}
+
+        for field_name, field_type in get_type_hints(
+            pyrdl_def, include_extras=True
+        ).items():
+            if field_name == "name" or field_name == "parameters":
+                continue
+
+            try:
+                constraint = irdl_to_attr_constraint(field_type, allow_type_var=True)
+            except TypeError as e:
+                raise PyRDLAttrDefinitionError(
+                    f"Invalid field type {field_type} for field name {field_name}."
+                ) from e
+
+            parameters[field_name] = constraint
+
         for field_name, value in clsdict.items():
             if field_name == "name":
+                # Ignore name field
                 continue
             if isinstance(
                 value, FunctionType | PropertyType | classmethod | staticmethod
             ):
+                # Ignore functions
                 continue
+
+            # Parameter def must be a field def
+            if isinstance(value, _ParameterDef):
+                if field_name not in parameters:
+                    raise PyRDLAttrDefinitionError(
+                        f"Missing field type for parameter name {field_name}"
+                    )
+
+                try:
+                    if value.param is not None:
+                        constraint = irdl_to_attr_constraint(
+                            value.param, allow_type_var=True
+                        )
+                        parameters[field_name] &= constraint
+                except TypeError as e:
+                    raise PyRDLAttrDefinitionError(
+                        f"Invalid constraint {value.param} for field name {field_name}."
+                    ) from e
+
+                continue
+
             # Constraint variables are allowed
```

Comment:
Should that still exist? Is there a case where we still have a value that is an `Annotated`?

Reply:
This hasn't got rid of all the Annotated, think the idea is to deprecate it in stages after this.